src/Pagination/Connection.hack

```diff
@@ -0,0 +1,196 @@
+namespace Slack\GraphQL\Pagination;
+
+use namespace HH\Lib\{C, Vec};
+use namespace Slack\GraphQL;
+
+/**
+ * Arguments passed to a connection's `paginate` method. These should be used to determine the result set.
+ *
+ * The framework ensures that at most one of `first` and `last` is provided.
+ */
+type PaginationArgs = shape(
+
+    /**
+     * If provided, only retrieve items which come before this cursor in the dataset.
+     */
+    ?'before' => string,
+
+    /**
+     * If provided, only retrieve items which come after this cursor in the dataset.
+     */
+    ?'after' => string,
+
+    /**
+     * If provided, only retrieve the first `first` items which appear after the `after` cursor or,
+     * if the `after` cursor was not provided, after the first item in the dataset.
+     */
+    ?'first' => int,
+
+    /**
+     * If provided, only retrieve the last `last` items which appear before the `before` cursor or,
+     * if the `before` cursor was not provided, before the last item in the dataset.
+     */
+    ?'last' => int,
+);
+
+/**
+ * `Connection` provides a standard mechanism for slicing and paginating querysets, as well as for providing cursors
+ * and informing the client when more results are available.
+ *
+ * Each connection paginates over a generic type `T`. `T` must be a Hack type which has a GraphQL representation that
+ * is not a list type. To paginate over a type `T`, subclass `Connection` and implement  `paginate` method. While
+ * `Connection` is purposely simple so as to support a wide variety of data sources, you're encouraged to create
+ * reusable subclasses which connect to the data sources you use.
+ *
+ * @see https://relay.dev/graphql/connections.htm for more information about GraphQL pagination.
+ * @see src/playground/UserConnection.hack for an example.
+ */
+abstract class Connection<T> {
+
+    /**
+     * Paginate the data set per the pagination args, which contain cursors and limits.
+     *
+     * For example, if the pagination args contain a `before` cursor with value `foo` and a `last` limit with value 5, 
+     * `paginate` should return edges wrapping the 5 items immediately prior to the item with cursor `foo`.
+     */
+    abstract protected function paginate(PaginationArgs $args): Awaitable<vec<Edge<T>>>;
```

Comment:
I decided to diverge from my spec and make the implementation construct each edge, rather than having the framework do this for them. This is to allow edges which have cursors that are not derived from a node's attribute but from some other data point, like the node's index in a list.